src/runtime/vm/vm.cc

```diff
@@ -521,13 +603,25 @@ int64_t VirtualMachine::LoadScalarInt(Index r) const {
   return result;
 }
 
-void VirtualMachine::RunLoop() {
+Index VirtualMachine::GetResultRegisterIndex() const {
+  Index op_ind = 0;
+  Instruction instr;
+  // TODO(vvchernov): can it be endless loop?
+  do {
+    instr = code_[op_ind++];
+  } while (instr.op == Opcode::Ret);
+
+  return instr.result;
+}
+
+void VirtualMachine::RunLoop(bool set_output_enabled) {
   ICHECK(this->exec_);
   ICHECK(this->code_);
   pc_ = 0;
   Index frame_start = frames_.size();
-  while (true) {
-  main_loop:
+  Index res_reg_index = GetResultRegisterIndex();
```

Comment:
For tuple results we'd need to redirect the allocs for the tuple indexes, which may be scattered quite widely within the code. So that suggests the compiler should record metadata for all that.

But I'm wondering if it would be better to bite-the-bullet and switch the VM to DPS. After that only Invoke-without-outputs would be the special case, requiring inspection of metadata to alloc the output tensors, make the call, and optionally construct the tuple result. I know that's a much bigger change, and I guess we could make that change after this PR given the invoke APIs will be the same.

Is there a high pressure customer use case which justifies that two step approach?

Reply:
As we can see from RunLoop design, and my tests with multiple outputs showed the same, only one ObjectRef (e.g. ADT) is return as result. It means we need only one index. Nevertheless tests are in progress to check possible issues. I'm not sure that current implementation is ready to merge.
@tmoreau89 what do you think there are any deadlines for this feature?

Reply:
Hello @mbs-octoml! I've updated my code for tuple results. I get register indices from AllocADT instruction and further there is the same action as for AllocTensor